### PR TITLE
Generation pipeline (first PR)

### PR DESCRIPTION
**Pull request by saridormi**

### CODE REVIEW

File: tests/test_data_processor.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+@pytest.mark.parametrize(
+    "test_input,expected_output",
+    [
+        (
+            "<FILE> some/path <nl> - smth.old{} <nl> + smth.new() <nl> unchanged line",
+            "some / path \n - smth . old { } \n + smth . new ( ) \n",
+        ),
+        (
+            "Binary files some/path/a and some/path/b differ <nl>",
+            "Binary files some / path / a and some / path / b differ \n",
```
Comment (SokolovYaroslav) [inline tests/test_data_processor.py:27]: А почему ожидается такое?) Ну мне просто интересно почему именно так

Comment (SokolovYaroslav) [reply tests/test_data_processor.py:27]: В следующем тесте понятно почему)

File: src/model/gpt2_decoder.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+        Note: signature is almost the same as in `generate` method from transformers,
+        see `transformers.generation_utils.GenerationMixin.generate` for more information.
+
+        :return: dictionary (with keys `sequences` and `scores`)
+        """
+        # use default values from model config for several parameters if they are not passed explicitly
+        num_beam_groups = num_beam_groups if num_beam_groups is not None else self.config.num_beam_groups
+        length_penalty = length_penalty if length_penalty is not None else self.config.length_penalty
+        early_stopping = early_stopping if early_stopping is not None else self.config.early_stopping
```
Comment (SokolovYaroslav) [inline src/model/gpt2_decoder.py:57]: Хмм, а кто такой `self.config` и откуда он берется?)

File: src/model/gpt2_decoder.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            logits_processor=logits_processors_list,
+            max_length=max_length,
+            pad_token_id=self.config.pad_token_id,
+            eos_token_id=self.config.eos_token_id,
+            output_scores=True,
+            return_dict_in_generate=True,
+            encoder_hidden_states=encoder_hidden_states,
+            encoder_attention_mask=encoder_attention_mask,
+        )
+        return {"sequences": beam_search_output.sequences, "scores": beam_search_output.sequences_scores}
```
Comment (SokolovYaroslav) [inline src/model/gpt2_decoder.py:123]: Конечно нужно внутрь смотреть, как именно он переиспользует выходы энкодера, пока не успел этого сделать(

**Review by SokolovYaroslav — COMMENTED** [review]
Пока не ясно, все ли прям хорошо переиспользуется, потому что надо внутрь бимсерча смотреть для этого.
Есть явное место для оптимизации -- пробелы между знаками препинания и не только. Они занимают места и наверное не так уж и важны. Можно попробовать в будущем обучить модель без них)
Завтра постараюсь посмотреть вглубь сорцов бимсерча

Ну и да, надо бы как-то переиспользовать один и тот же препроцессор в инференсе и обучении. Могу сказать по своему опыту, что это прям важно)

А так выглядит хорошо)

File: .gitignore [hunk 4]
```diff
@@ -1,4 +1,5 @@
-/raw_data/
+data/
+raw_data/
 __pycache__/
 .idea
 /asd.py
```
[the remark this answers is not in the document]
Comment (SokolovYaroslav) [reply .gitignore:5]: Ну я про `/asd.py`

[the remark this answers is not in the document]
Comment (saridormi) [reply .gitignore:5]: ой, это кажется еще с основной ветки осталось
я иногда в таком файле писала что-то рандомное, чтобы разобраться, правильно ли я понимаю, как что-нибудь работает в PyTorch/PyTorch Lightning/Transformers :upside_down_face: думаю могу убрать

File: README.md [hunk 5]
```diff
@@ -1,14 +1,105 @@
...
+```
+data_processor:
+  arg: ...
+model:
+  arg: ...
+generation_kwargs:
+  arg: ...
+device: ...
+min_length: ...
+max_length: ...
```
[the remark this answers is not in the document]
Comment (saridormi) [reply README.md]: да, согласна, поправила все в README  :)

File: src/__init__.py [hunk 6]
```diff
@@ -0,0 +1,3 @@
+from . import data_utils, model
+
+__all__ = ["data_utils", "model"]
```
[the remark this answers is not in the document]
Comment (saridormi) [reply src/__init__.py:3]: попробовала убрать это, ничего не сломалось, так что правда
тогда оставлю пустой файл

File: src/data_utils/data_processor.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        # add <bos> and <eos> tokens
+        msg = torch.cat(
+            (
+                torch.ones(1, 1) * self.msg_tokenizer.bos_token_id,
+                msg,
+                torch.ones(1, 1) * self.msg_tokenizer.eos_token_id,
+            ),
+            dim=1,
+        )
+        return msg.long()
```
[the remark this answers is not in the document]
Comment (saridormi) [reply src/data_utils/data_processor.py:204]: это именно про конкатенацию или про весь препроцессинг?

[the remark this answers is not in the document]
Comment (saridormi) [reply src/data_utils/data_processor.py:204]: вообще я решила что-то немного поменять, потому что мне показалось, что в коде для обучения много какой-то не нужной здесь функциональности, связанной с построением истории, записью в файлы и т.д.
а код для конкатенации вообще вынесен в отдельный класс, чтобы его передавать в DataLoader
но я вроде постаралась, чтобы происходило то же самое

File: src/data_utils/data_processor.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        diff_tokenizer_name_or_path: str,
+        msg_tokenizer_name_or_path: str,
+        preprocessing: bool = False,
+        nl_token: str = "\n",
+    ):
+        self.history_max_len = history_max_len
+        self.prompt_max_len = prompt_max_len
+        self.diff_tokenizer = AutoTokenizer.from_pretrained(diff_tokenizer_name_or_path, use_fast=True)
+        self.msg_tokenizer = AutoTokenizer.from_pretrained(msg_tokenizer_name_or_path, use_fast=True)
+        self.nl_token = nl_token  # this might not be needed? (newline char in input is most likely \n by default)
```
[the remark this answers is not in the document]
Comment (saridormi) [reply src/data_utils/data_processor.py:27]: он используется в основном для разделения диффов на строки, чтобы можно было легко различать неизмененные или всякие служебные строки с теми, которые мы хотим оставить
то есть, если в данных по каким-то причинам используется не `\n`, а что-то еще, можно сюда передать этот токен
мне это почти сразу показалось странным, но пока решила оставить, например, в датасете у нас везде `<nl>`

File: src/model/gpt2_decoder.py (hunk 2 above)
Comment (SokolovYaroslav) [reply src/model/gpt2_decoder.py:57]: Ну кто такой конфиг понятно, да
Хмм, а эти поля у него по умолчанию просто есть, да?

File: src/model/gpt2_decoder.py (hunk 3 above)
Comment (SokolovYaroslav) [reply src/model/gpt2_decoder.py:123]: Хмм, ну выглядит так, что все переиспользуется как надо

File: src/data_utils/data_processor.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        """
+        # in case with empty list just return empty tensor
+        if not inputs:
+            return torch.empty(0)
+        tokenized_inputs = tokenizer(
+            inputs,
+            padding=False,
+            truncation=True,
+            return_tensors="pt",
+            max_length=500,
```
[the remark this answers is not in the document]
Comment (saridormi) [reply src/data_utils/data_processor.py]: кажется, в transformers нельзя задать, с какой стороны хочется обрезать входные данные, поэтому пока сделала так: параметры `truncation=True, max_length=500` передаются только в случае с диффами, а сообщения [обрезаются слева перед конкатенацией с историей](https://github.com/JetBrains-Research/commit_message_generation/blob/38a00f404d8534148e1956bf96dbc104845548cc/src/data_utils/data_processor.py#L173)

File: src/model/gpt2_decoder.py (hunk 2 above)
Comment (saridormi) [reply src/model/gpt2_decoder.py:57]: да, это вроде у всех моделей из Transformers есть по умолчанию
